model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java

```diff
@@ -700,54 +702,80 @@ public int getUsersCount(RealmModel realm) {
-                parameterName = JpaUserProvider.EMAIL;
+            String key = entry.getKey();
+            String value = entry.getValue();
+
+            if (value == null) {
+                continue;
+            }
+
+            value = value.toLowerCase();
+
+            if (key.equals(UserModel.USERNAME)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.USERNAME)), "%" + value + "%"));
+            } else if (key.equalsIgnoreCase(UserModel.FIRST_NAME)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.FIRST_NAME)), "%" + value + "%"));
+            } else if (key.equalsIgnoreCase(UserModel.LAST_NAME)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.LAST_NAME)), "%" + value + "%"));
+            } else if (key.equalsIgnoreCase(UserModel.EMAIL)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.EMAIL)), "%" + value + "%"));
+            } else if (UserModel.GROUPS.equals(key)) {
+                Subquery<String> subquery = queryBuilder.subquery(String.class);
+                Root<UserGroupMembershipEntity> from = subquery.from(UserGroupMembershipEntity.class);
+                Join<Object, Object> user1 = from.join("user");
+                Join<Object, Object> user = user1.on(builder.equal(user1.get("id"), root.get("id")));
+                subquery.select(user.get("id"));
+                List<Predicate> subPredicates = new ArrayList<>();
+
+                try {
+                    List list = JsonSerialization.readValue(value, List.class);
+
+                    for (Object o : list) {
+                        subPredicates.add(builder.equal(from.get("groupId"), o));
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+
+                subquery.where(subPredicates.toArray(new Predicate[subPredicates.size()]));
+
+                predicates.add(root.get("id").in(subquery));
             }
-            if (parameterName == null) continue;
-            query.setParameter(parameterName, "%" + entry.getValue().toLowerCase() + "%");
         }
+
+        queryBuilder.where(predicates.toArray(new Predicate[predicates.size()])).orderBy(builder.asc(root.get(UserModel.USERNAME)));
+
+        Query query = em.createQuery(queryBuilder);
+
         if (firstResult != -1) {
             query.setFirstResult(firstResult);
         }
+
         if (maxResults != -1) {
             query.setMaxResults(maxResults);
         }
-        List<UserEntity> results = query.getResultList();
-        List<UserModel> users = new ArrayList<UserModel>();
-        for (UserEntity entity : results) users.add(new UserAdapter(session, realm, em, entity));
+
+        List<String> results = query.getResultList();
+        List<UserModel> users = new ArrayList<>();
+
+        for (String id : results) {
+            users.add(getUserById(id, realm));
```

Comment:
This does look like an n+1 query to me. Wouldn't it be possible to fetch the users directly inside the above query?